src/ast/ast.c

```diff
@@ -0,0 +1,328 @@
+        const cypher_astnode_t *child = cypher_ast_query_get_clause(ast->root, i);
+        if (cypher_astnode_type(child) != type) continue;
+
+        found = array_append(found, child);
+    }
+
+    return found;
+}
+
+void _AST_CollectAliases(const char ***aliases, const cypher_astnode_t *entity) {
+    if (entity == NULL) return;
+
+    if (cypher_astnode_type(entity) == CYPHER_AST_IDENTIFIER) {
+        const char *identifier = cypher_ast_identifier_get_name(entity);
+        *aliases = array_append(*aliases, identifier);
+        return;
+    }
+
+    uint nchildren = cypher_astnode_nchildren(entity);
+    for (uint i = 0; i < nchildren; i ++) {
+        _AST_CollectAliases(aliases, cypher_astnode_get_child(entity, i));
+    }
+}
+
+// Collect aliases from clauses that introduce entities (MATCH, MERGE, CREATE, UNWIND)
+const char** AST_CollectAliases(AST *ast) {
+    const char **aliases = array_new(const char*, 1);
+    uint clause_count = cypher_ast_query_nclauses(ast->root);
+    for (uint i = 0; i < clause_count; i ++) {
+        const cypher_astnode_t *clause = cypher_ast_query_get_clause(ast->root, i);
+        cypher_astnode_type_t type = cypher_astnode_type(clause);
+        if (type == CYPHER_AST_MATCH  ||
+            type == CYPHER_AST_MERGE  ||
+            type == CYPHER_AST_CREATE ||
+            type == CYPHER_AST_UNWIND
+            ) {
+            _AST_CollectAliases(&aliases, clause);
+        }
+    }
+
+    return aliases;
+}
+
+const cypher_astnode_t* AST_GetBody(const cypher_parse_result_t *result) {
+    const cypher_astnode_t *statement = cypher_parse_result_get_root(result, 0);
+    assert(statement && cypher_astnode_type(statement) == CYPHER_AST_STATEMENT);
+
+    return cypher_ast_statement_get_body(statement);
+}
+
+AST* AST_Build(cypher_parse_result_t *parse_result) {
+    AST *ast = rm_malloc(sizeof(AST));
+    ast->entity_map = NULL;
+    ast->root = AST_GetBody(parse_result);
+    assert(ast->root);
+    if (cypher_astnode_type(ast->root) == CYPHER_AST_QUERY) AST_BuildEntityMap(ast);
+
+    // Set thread-local AST
+    pthread_setspecific(_tlsASTKey, ast);
+
+    return ast;
+}
+
+AST* AST_NewSegment(AST *master_ast, uint start_offset, uint end_offset) {
+    AST *ast = rm_malloc(sizeof(AST));
+
+    uint n = end_offset - start_offset;
+
+    cypher_astnode_t *clauses[n];
+    for (uint i = 0; i < n; i ++) {
+        clauses[i] = (cypher_astnode_t*)cypher_ast_query_get_clause(master_ast->root, i + start_offset);
+    }
+    struct cypher_input_range range = {};
+    ast->root = cypher_ast_query(NULL, 0, (cypher_astnode_t *const *)clauses, n, NULL, 0, range);
+
+    pthread_setspecific(_tlsASTKey, ast); // TODO I don't know if I like this
+    AST_BuildEntityMap(ast);
+
+    return ast;
+}
+
+// TODO I find it so weird that this is necessary
+long AST_ParseIntegerNode(const cypher_astnode_t *int_node) {
+    assert(int_node);
+
+    const char *value_str = cypher_ast_integer_get_valuestr(int_node);
```

Comment:
Numeric values are represented by the parser internally as strings?

Reply:
Yeah. My guess is that Leishman didn't want the parser to be responsible for numerics that are smaller or larger than the range of standard integer types, but it's definitely a nuisance for us.